File: go/conformance/matchers.go
package conformance

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

// Captures holds $capture:name values for one case. It is what makes G4
// expressible: a fresh Idempotency-Key per logical message, the same one on
// that message's retry.
type Captures map[string]string

var matcherPattern = regexp.MustCompile(`^\$(any|absent|capture|same|notSame)(?::(.+))?$`)

func isEmpty(v any) bool {
	if v == nil {
		return true
	}
	s, ok := v.(string)
	return ok && s == ""
}

// matchValue reports an empty string on success or a reason on failure.
//
// $absent is NOT handled here: absence is a property of the parent object, so
// matchPartial deals with it before ever reaching a value.
func matchValue(expected, actual any, captures Captures, path string) string {
	if s, ok := expected.(string); ok {
		if m := matcherPattern.FindStringSubmatch(s); m != nil {
			kind, name := m[1], m[2]
			switch kind {
			case "any":
				if isEmpty(actual) {
					return fmt.Sprintf("%s: expected any non-empty value, got %s", path, format(actual))
				}
				return ""
			case "absent":
				return fmt.Sprintf("%s: $absent is only valid as an object field, not as a value", path)
			case "capture":
				if isEmpty(actual) {
					return fmt.Sprintf("%s: cannot capture an empty value", path)
				}
				got := format(actual)
				if prior, seen := captures[name]; seen && prior != got {
					return fmt.Sprintf("%s: %s was already captured as %s, now saw %s", path, name, prior, got)
				}
				captures[name] = got
				return ""
			case "same":
				prior, seen := captures[name]
				if !seen {
					return fmt.Sprintf("%s: $same:%s used before %s was captured", path, name, name)
				}
				if format(actual) != prior {
					return fmt.Sprintf("%s: expected the captured %s (%s), got %s", path, name, prior, format(actual))
				}
				return ""
			case "notSame":
				prior, seen := captures[name]
				if !seen {
					return fmt.Sprintf("%s: $notSame:%s used before %s was captured", path, name, name)
				}
				if isEmpty(actual) {
					return fmt.Sprintf("%s: expected a non-empty value differing from %s, got %s", path, name, format(actual))
				}
				if format(actual) == prior {
					return fmt.Sprintf("%s: expected a value DIFFERENT from %s, but got the same one (%s)", path, name, format(actual))
				}
				return ""
			}
		}
	}

	switch want := expected.(type) {
	case map[string]any:
		return matchPartial(want, actual, captures, path)
	case []any:
		got, ok := actual.([]any)
		if !ok {
			return fmt.Sprintf("%s: expected an array, got %s", path, format(actual))
		}
		if len(want) != len(got) {
			return fmt.Sprintf("%s: expected %d elements, got %d", path, len(want), len(got))
		}
		for i := range want {
			if reason := matchValue(want[i], got[i], captures, fmt.Sprintf("%s[%d]", path, i)); reason != "" {
				return reason
			}
		}
		return ""
	}

	if format(expected) != format(actual) {
		return fmt.Sprintf("%s: expected %s, got %s", path, format(expected), format(actual))
	}
	return ""
}

// matchPartial compares only the declared fields, ignoring the rest. That is
// deliberate: a case should pin what it is about and stay silent on everything
// else, so unrelated payload growth does not break every case at once.
func matchPartial(expected map[string]any, actual any, captures Captures, path string) string {
	target, ok := actual.(map[string]any)
	if !ok {
		return fmt.Sprintf("%s: expected an object, got %s", path, format(actual))
	}
	keys := sortedKeys(expected)
	for _, key := range keys {
		at := key
		if path != "" {
			at = path + "." + key
		}
		want := expected[key]
		if s, isString := want.(string); isString && s == "$absent" {
			// Presence of the KEY is the failure. A decoded JSON null arrives
			// here as nil, and treating that as absent would accept the exact
			// mistake m1 exists to catch.
			if got, present := target[key]; present {
				return fmt.Sprintf("%s: expected the field to be absent, but it was %s", at, format(got))
			}
			continue
		}
		got, present := target[key]
		if !present {
			return at + ": missing"
		}
		if reason := matchValue(want, got, captures, at); reason != "" {
			return reason
		}
	}
	return ""
}

// matchHeaders compares case-insensitively by name.
func matchHeaders(expected map[string]string, actual map[string]string, captures Captures) string {
	lowered := map[string]any{}
	for k, v := range actual {
		lowered[strings.ToLower(k)] = v
	}
	wanted := map[string]any{}
	for k, v := range expected {
		wanted[strings.ToLower(k)] = v
	}
	return matchPartial(wanted, lowered, captures, "headers")
}

// format renders a value the same way regardless of how JSON decoded it, so a
// number that arrived as float64 compares equal to the same number in a case.
func format(v any) string {
	switch value := v.(type) {
	case nil:
		return "null"
	case string:
		return value
	case float64:
		if value == float64(int64(value)) {
			return fmt.Sprintf("%d", int64(value))
		}
		return fmt.Sprintf("%v", value)
	default:
		encoded, err := json.Marshal(value)
		if err != nil {
			return fmt.Sprintf("%v", value)
		}
		return string(encoded)
	}
}

func sortedKeys(m map[string]any) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	// Deterministic order: Go randomises map iteration, and a runner whose
	// failure list reorders between runs is miserable to read.
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j] < keys[j-1]; j-- {
			keys[j], keys[j-1] = keys[j-1], keys[j]
		}
	}
	return keys
}

File: python/conformance/matchers.py
"""Value matching for conformance cases. Mirrors ../../conformance/SCHEMA.md."""

from __future__ import annotations

import json
import re
from typing import Any

Captures = dict[str, str]

_MATCHER = re.compile(r"^\$(any|absent|capture|same|notSame)(?::(.+))?$")


def _is_empty(value: Any) -> bool:
    return value is None or value == ""


def format_value(value: Any) -> str:
    """Render a value the same way regardless of how JSON decoded it.

    A number that arrived as ``float`` must compare equal to the same number
    written as an int in a case.
    """
    if value is None:
        return "null"
    if isinstance(value, str):
        return value
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(int(value)) if float(value).is_integer() else str(value)
    return json.dumps(value, sort_keys=True)


def match_value(expected: Any, actual: Any, captures: Captures, path: str) -> str:
    """Return an empty string on success or a reason on failure.

    ``$absent`` is NOT handled here: absence is a property of the parent
    object, so ``match_partial`` deals with it before reaching a value.
    """
    if isinstance(expected, str):
        matcher = _MATCHER.match(expected)
        if matcher:
            kind, name = matcher.group(1), matcher.group(2)
            if kind == "any":
                return "" if not _is_empty(actual) else (
                    f"{path}: expected any non-empty value, got {format_value(actual)}"
                )
            if kind == "absent":
                return f"{path}: $absent is only valid as an object field, not as a value"
            if kind == "capture":
                if _is_empty(actual):
                    return f"{path}: cannot capture an empty value"
                got = format_value(actual)
                prior = captures.get(name)
                if prior is not None and prior != got:
                    return f"{path}: {name} was already captured as {prior}, now saw {got}"
                captures[name] = got
                return ""
            if kind == "same":
                if name not in captures:
                    return f"{path}: $same:{name} used before {name} was captured"
                if format_value(actual) != captures[name]:
                    return (
                        f"{path}: expected the captured {name} ({captures[name]}), "
                        f"got {format_value(actual)}"
                    )
                return ""
            if kind == "notSame":
                if name not in captures:
                    return f"{path}: $notSame:{name} used before {name} was captured"
                if _is_empty(actual):
                    return (
                        f"{path}: expected a non-empty value differing from {name}, "
                        f"got {format_value(actual)}"
                    )
                if format_value(actual) == captures[name]:
                    return (
                        f"{path}: expected a value DIFFERENT from {name}, but got the "
                        f"same one ({format_value(actual)})"
                    )
                return ""

    if isinstance(expected, dict):
        return match_partial(expected, actual, captures, path)
    if isinstance(expected, list):
        if not isinstance(actual, list):
            return f"{path}: expected an array, got {format_value(actual)}"
        if len(expected) != len(actual):
            return f"{path}: expected {len(expected)} elements, got {len(actual)}"
        for index, item in enumerate(expected):
            reason = match_value(item, actual[index], captures, f"{path}[{index}]")
            if reason:
                return reason
        return ""

    if format_value(expected) != format_value(actual):
        return f"{path}: expected {format_value(expected)}, got {format_value(actual)}"
    return ""


def match_partial(expected: dict[str, Any], actual: Any, captures: Captures, path: str) -> str:
    """Compare only the declared fields, ignoring the rest.

    Deliberate: a case should pin what it is about and stay silent on
    everything else, so unrelated payload growth does not break every case.
    """
    if not isinstance(actual, dict):
        return f"{path}: expected an object, got {format_value(actual)}"
    for key in sorted(expected):
        at = f"{path}.{key}" if path else key
        want = expected[key]
        if want == "$absent":
            # Presence of the KEY is the failure. A decoded JSON null arrives
            # here as None, and treating that as absent would accept the exact
            # mistake m1 exists to catch.
            if key in actual:
                return f"{at}: expected the field to be absent, but it was {format_value(actual[key])}"
            continue
        if key not in actual:
            return f"{at}: missing"
        reason = match_value(want, actual[key], captures, at)
        if reason:
            return reason
    return ""


def match_headers(expected: dict[str, str], actual: dict[str, str], captures: Captures) -> str:
    """Header names are case-insensitive."""
    lowered = {k.lower(): v for k, v in actual.items()}
    wanted = {k.lower(): v for k, v in expected.items()}
    return match_partial(wanted, lowered, captures, "headers")
